package buffers;

message Node {
	optional              uint32 id = 1;
	optional   uint64 time_register = 2;
	optional uint64 time_unregister = 3;
	required          string hwdesc = 4;
	required          string swdesc = 5;
	required          string indesc = 6;
}

message ProcessID {
	required  uint32 nid = 1;
	required  uint32 pid = 2; 
	required uint32 time = 3;
}


message ComponentID {
	required ProcessID pid = 1;
	required     uint32 id = 2;
}


message ActivityID {
	required ComponentID cid = 1;
	required       uint32 id = 2;
}


message AttributeValue {

	enum Type {
		INT64   = 0;
		UINT64  = 1;
		INT32   = 2;
		UINT32  = 3;
		FLOAT   = 4;
		DOUBLE  = 5;
		INVALID = 6;
		STRING  = 7;
	}
	
	required   Type type = 1;
	optional   int32 i32 = 2;
	optional   int64 i64 = 3;
	optional uint32 ui32 = 4;
	optional uint64 ui64 = 5;
	optional     float f = 6;
	optional    double d = 7;
	optional  string str = 8;

}


message Attribute {
	required            uint32 id = 1;
	required AttributeValue value = 2;
}


message RemoteCallIdentifier {
	optional      uint32 nid = 1;
	optional     uint32 uuid = 2;
	optional uint32 instance = 3;
}


message RemoteCall {
	required RemoteCallIdentifier target = 1;
	repeated        Attribute attributes = 2;
}


message Activity {
	required           ActivityID aid = 1;
	required            uint32 ucaid =  2;
	required       uint64 time_start =  3;
	optional        uint64 time_stop =  4;
	optional      uint32 error_value =  5;
	optional   uint64 remote_invoker =  6;
	repeated      ActivityID parents = 10;
	repeated    Attribute attributes = 11;
	repeated RemoteCall remote_calls = 12;
}


message MessageBuffer {

	enum Action {
		Advertise      = 0;
		Component      = 1;
		Subscribe      = 2;
		Activity       = 3;
		RegisterNode   = 4;
		UnregisterNode = 5;
		Log            = 6;
	}
	
	required   Action action = 1 [default = Activity];
	required     uint32 type = 2;
	optional uint32 priority = 3;

	optional         Node node = 10;
	optional Activity activity = 11;
	optional     string logmsg = 12;
}

